### ISSUE TITLE
False dead_code warning on struct pattern match

### ISSUE THREAD

**Issue by nvzqz** [rust-lang/rust#56750]
The following code will emit a `dead_code` warning ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5c94d6fd6a0bf1512d68495aedbd1b82)):

```rust
use std::mem;

fn main() {
    struct Value {
        a: i32,
        b: i32,
    };

    let Value { a, b } = unsafe { mem::zeroed() };
    println!("{} {}", a, b);
}
```

It states that `Value` is never constructed when in fact it has been within the `let Value { a, b }` pattern match.

```
   Compiling playground v0.0.1 (/playground)
warning: struct is never constructed: `Value`
 --> src/main.rs:4:5
  |
4 |     struct Value {
  |     ^^^^^^^^^^^^
  |
  = note: #[warn(dead_code)] on by default

    Finished dev [unoptimized + debuginfo] target(s) in 0.90s
     Running `target/debug/playground`
```

**Comment by zackmdavis** [comment]
@nvzqz Thanks for the report! We recently [changed the wording](https://github.com/rust-lang/rust/pull/52332) to say "never constructed" rather than "never used" in the hopes of clarifying that the [liveness analysis](https://github.com/rust-lang/rust/blob/dd8fc7dc06dea00afbd365468cf4804f68a3531c/src/librustc/middle/liveness.rs) doesn't consider destructuring patterns as "live," but the existence of this Issue would seem to suggest that this still isn't clear enough. :slightly_frowning_face: 

Part of me was tempted to call this #wontfix (as it wasn't obvious to me why it's useful to have a struct pattern without the struct being constructed anywhere else—what would realistically be on the other side of the match/assignment?), but your [linked commit](https://github.com/oceanpkg/os-utils/commit/5224332a5fc) being a real-world counterexample makes me more sympathetic to changing the analysis to consider patterns "live." (As you point out, `unsafe` bytes can be on the other side of the match, and you might want to have a struct to name them.)

**Comment by nvzqz** [comment]
I noticed the warning for `Value` not being constructed but in my reasoning, it actually is when returned from `mem::zeroed()` and then immediately deconstructed in the pattern. I know that's a very implicit way of viewing it. I still think that code like this that's correct shouldn't emit warnings despite the warning being generated in `unsafe` code.

Thanks for taking my real-world case in that commit into account!

**Comment by danielhenrymantilla** [comment]
Using a struct only for it to be destructured right away does indeed have its use cases, see also: https://github.com/danielhenrymantilla/fstrings-rs/blob/2de1a3db8d6bf6f9e33ea6eeef8468d75b329c83/src/proc_macro/mod.rs#L91-L94:

```rust
let Args {
    format_literal,
    mut extra_args,
} = parse_macro_input!(input); // Args : Parse
```

  - My `Args` struct implements `Parse` logic, but then I just want to extract its fields right away.

  - Thus my `Args` struct is definitely not `dead_code`.

**Comment by Spoonbender** [comment]
Triage: no change

**Comment by mu001999** [comment]
This has been fixed by https://github.com/rust-lang/rust/pull/142485.
